### PR TITLE
chore: destination logic refactor

### PR DESCRIPTION
**Pull request by UncleSamtoshi**
- refactor destination logic so that it can be shared across screens
- enable parsing of lnurlwithdraw strings from the SendBitcoinDestinationScreen

### CODE REVIEW

File: app/screens/send-bitcoin-screen/send-bitcoin-details-screen.tsx [hunk 1]
```diff
@@ -256,7 +254,9 @@ const SendBitcoinDetailsScreen = ({
   const [isModalVisible, setIsModalVisible] = useState(false)
   const [errorMessage, setErrorMessage] = useState("")
   const [validAmount, setValidAmount] = useState(false)
-  const usdDisabled = paymentType === "onchain" || usdWalletId === undefined
+  const usdDisabled =
```
Comment (nicolasburtey) [inline app/screens/send-bitcoin-screen/send-bitcoin-details-screen.tsx:257]: I feel we could remove this assumption for now. I don't think we need this complexity of managing having or not having usd enabled. can be done as part of another PR.

File: app/screens/send-bitcoin-screen/destination-information.tsx [hunk 2]
```diff
@@ -6,21 +6,22 @@ import { useAppConfig } from "@app/hooks"
...
 import { IntraledgerPaymentDestination } from "@galoymoney/client/dist/parsing-v2"
+import { InvalidDestinationReason } from "./payment-destination/index.types"
 
 const toLnAddress = (handle: string, lnDomain: string) => {
   return `${handle}@${lnDomain}`
 }
 
 const destinationStateToInformation = (
-  destinationState: SendBitcoinDestinationState,
+  sendBitcoinReducerState: SendBitcoinDestinationState,
```
Comment (nicolasburtey) [inline app/screens/send-bitcoin-screen/destination-information.tsx:18]: I think we should also keep, for consistency, the pattern we have in the backend of using object are argument, instead of positional argument. 

we can do those changes in another PR

File: app/screens/send-bitcoin-screen/destination-information.tsx [hunk 3]
```diff
@@ -6,21 +6,22 @@ import { useAppConfig } from "@app/hooks"
...
   translate: TranslationFunctions,
```
Comment (nicolasburtey) [inline app/screens/send-bitcoin-screen/destination-information.tsx:19]: you could just pass `translate.SendBitcoinDestinationScreen`. I think otherwise we typically call the translate function `TT` in the other part of the app (assuming this is the same thing). we should keep the same function name here

File: app/screens/send-bitcoin-screen/destination-information.tsx [hunk 4]
```diff
@@ -6,21 +6,22 @@ import { useAppConfig } from "@app/hooks"
...
 import { Text, View } from "react-native"
 import EStyleSheet from "react-native-extended-stylesheet"
-import { SendBitcoinDestinationState } from "./send-bitcoin-reducer"
+import { DestinationState, SendBitcoinDestinationState } from "./send-bitcoin-reducer"
 import { IntraledgerPaymentDestination } from "@galoymoney/client/dist/parsing-v2"
+import { InvalidDestinationReason } from "./payment-destination/index.types"
 
 const toLnAddress = (handle: string, lnDomain: string) => {
   return `${handle}@${lnDomain}`
 }
```
Comment (nicolasburtey) [inline app/screens/send-bitcoin-screen/destination-information.tsx:15]: lnDomain doesn't change. this could be memoized, such that when it's called only the handle needs to be passed 

```suggestion
const toLnAddress = (lnDomain: string)(handle: string)  => {
  return `${handle}@${lnDomain}`
}
```

Comment (nicolasburtey) [reply app/screens/send-bitcoin-screen/destination-information.tsx:15]: means `toLnAddress ` should be in `destinationStateToInformation`

File: app/screens/send-bitcoin-screen/scanning-qrcode-screen.tsx [hunk 5]
```diff
@@ -93,33 +87,29 @@ type ScanningQRCodeScreenProps = {
...
+      }
+      contacts {
+        id
+        username
+      }
+    }
+  }
+
+  # TODO replace with AccountDefaultWallet?
+  query userDefaultWalletId($username: Username!) {
```
Comment (nicolasburtey) [inline app/screens/send-bitcoin-screen/scanning-qrcode-screen.tsx:111]: I didn't know we could copy the query multiple times

File: app/screens/send-bitcoin-screen/send-bitcoin-destination-screen.tsx [hunk 6]
```diff
@@ -213,232 +161,87 @@ const SendBitcoinDestinationScreen = ({
...
-    return async (username: string) =>
-      getUsernameStatus({
-        username,
-        contacts: lowercaseContacts,
-        myUsername: lowercaseMyUsername,
-        getWalletIdForUsername,
-      })
-  }, [contacts, userDefaultWalletIdQuery, myUsername])
+  const [userDefaultWalletIdQuery] = useUserDefaultWalletIdLazyQuery({
+    fetchPolicy: "cache-and-network",
```
Comment (nicolasburtey) [inline app/screens/send-bitcoin-screen/send-bitcoin-destination-screen.tsx:170]: it's a lazy query. I don't think we want to use `cache and network`:
- because its a lazy query, it acts like a mutation and will only be called when the function is called instead of when the screen is loaded. 

so I'm not sure how we would get the network query backed from this function if the value is cached. I'll assume we don't

also, we may not want to cache this at all, so maybe `no-cache` is the appropriate solution.

File: app/screens/send-bitcoin-screen/payment-destination/lnurl.ts [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+
+export type ResolveLnurlDestinationParams = {
+  parsedLnurlDestination: LnurlPaymentDestination
+  lnurlDomains: string[]
+  userDefaultWalletIdQuery: LazyQueryExecFunction<
+    UserDefaultWalletIdQuery,
+    Exact<{
+      username: string
+    }>
+  >
```
Comment (nicolasburtey) [inline app/screens/send-bitcoin-screen/payment-destination/lnurl.ts:27]:
```suggestion
  userDefaultWalletIdQuery: UserDefaultWalletIdLazyQueryHookResult
```

Comment (nicolasburtey) [reply app/screens/send-bitcoin-screen/payment-destination/lnurl.ts:27]: there are many instance where the type that could be simplify with the diff above

File: app/screens/send-bitcoin-screen/payment-destination/index.types.ts [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+  rawInput: string
+  myWalletIds: string[]
+  bitcoinNetwork: Network
+  lnurlDomains: string[]
+  userDefaultWalletIdQuery: LazyQueryExecFunction<
+    UserDefaultWalletIdQuery,
+    Exact<{
+      username: string
+    }>
+  >
```
Comment (nicolasburtey) [inline app/screens/send-bitcoin-screen/payment-destination/index.types.ts:31]:
```suggestion
  userDefaultWalletIdQuery: UserDefaultWalletIdLazyQueryHookResult[0]
```

File: app/screens/send-bitcoin-screen/payment-destination/index.types.ts [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+}
+
+export type ValidParsedReceiveDestination = LnurlWithdrawDestination
+
+export type LnurlWithdrawDestination = {
+  valid: true
+  lnurl: string
+  callback: string
+  domain: string
+  k1: string
```
Comment (nicolasburtey) [inline app/screens/send-bitcoin-screen/payment-destination/index.types.ts:109]: what is `k1`?

File: app/screens/send-bitcoin-screen/send-bitcoin-destination-screen.tsx (hunk 6 above)
Comment (UncleSamtoshi) [reply app/screens/send-bitcoin-screen/send-bitcoin-destination-screen.tsx:170]: Yeah my one thinking for cache and network was that the call would always succeed instantly if it was in the cache, and then the next time it was called it would be the up to date value. The trade off would be that if someone changed default wallets, there would be one payment to them using the old wallet id. I suppose however that if the getWaleltId query fails there will probably be other things failing and it wont be worth it to try and prevent that situation.

Comment (nicolasburtey) [reply app/screens/send-bitcoin-screen/send-bitcoin-destination-screen.tsx:170]: >  I suppose however that if the getWaleltId query fails there will probably be other things failing and it wont be worth it to try and prevent that situation.

exactly. 

I think this is like the getFee where we try to preemptively calculate them locally if this is from our node instead of calling the server. the benefits do not outweigh the added complexity

File: app/screens/send-bitcoin-screen/destination-information.tsx (hunk 3 above)
Comment (UncleSamtoshi) [reply app/screens/send-bitcoin-screen/destination-information.tsx:19]: I think this should be handled in when creating the fix for issue #1089

File: app/screens/send-bitcoin-screen/payment-destination/index.types.ts (hunk 9 above)
Comment (UncleSamtoshi) [reply app/screens/send-bitcoin-screen/payment-destination/index.types.ts:109]: It is part of the lnurl withdraw spec that the lnurl server uses to identify the calling wallet https://github.com/lnurl/luds/blob/luds/03.md

File: app/screens/send-bitcoin-screen/payment-destination/index.types.ts (hunk 8 above)
Comment (UncleSamtoshi) [reply app/screens/send-bitcoin-screen/payment-destination/index.types.ts:31]: Changed it

File: app/screens/send-bitcoin-screen/destination-information.tsx (hunk 2 above)
Comment (UncleSamtoshi) [reply app/screens/send-bitcoin-screen/destination-information.tsx:18]: This can be addressed with issue #1089

**Review by nicolasburtey — APPROVED** [review]

**Review by nicolasburtey — APPROVED** [review]
